torch/_higher_order_ops/scan.py

```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+        joint_graph = ctx._joint_graph
+        dim = ctx._dim
+        num_leaves_init = ctx._num_leaves_init
+        num_leaves_xs = ctx._num_leaves_xs
+        reverse = ctx._reverse
+        num_leaves_ys = ctx._num_leaves_ys
+
+        # The results from the forward scan are always stacked on dim 0
+        # The gradients though need to be provided with the correct scan dimension dim
+        # Therefore, the inputs to the backward scan are all on dim 0, and the scan is performed on dim 0
+        # The gradient outputs are finally shifted at the end to the correct dim
+        bwd_scan_dim = 0
+
+        # Retrieve the forward inputs and the forward outputs
+        operands_outs = ctx.saved_tensors
+        init = operands_outs[:num_leaves_init]
+        # The backward scan operates on the 0-th dim and thus the original inputs need to be
+        # permuted accordingly
+        xs = [
+            shift_source_dim_to_target_dim(o, dim, bwd_scan_dim)
+            for o in operands_outs[num_leaves_init : num_leaves_init + num_leaves_xs]
+        ]
+        # The forward path stores init + xs + carries + additional_inputs
+        # Thus carries (of length num_leaves_init) start at positions num_leaves_init + num_leaves_xs
+        # and end with 2 * num_leaves_init + num_leaves_xs
+        carries = operands_outs[
+            num_leaves_init + num_leaves_xs : 2 * num_leaves_init + num_leaves_xs
+        ]
+        additional_inputs = list(operands_outs[2 * num_leaves_init + num_leaves_xs :])
+
+        with torch._C._AutoDispatchBelowAutograd():
+            # The flat gradients are a list of g_c_T, g_ys and optionally the gradients for the additional_inputs
+            flat_grads_list = list(flat_grads)
+            g_c_T = flat_grads_list[:num_leaves_init]
+            g_ys = flat_grads_list[num_leaves_init : num_leaves_init + num_leaves_ys]
+            old_g_additional_inputs_T = flat_grads_list[
+                num_leaves_init + num_leaves_ys :
+            ]
+
+            # Prepare the inputs for the backward scan.
+            # This involves flipping the input xs if needed as well as
+            # Prepending the init of the forward scan to the carries
+            xs, carries = prepare_xs_carries_for_bwd(
+                xs, init, carries, bwd_scan_dim, reverse
+            )
+            xs_bwd = [*g_ys, *carries, *xs]
+
+            g_init, g_xs = _extract_carry_and_out(
+                scan_op(
+                    joint_graph, g_c_T, xs_bwd, bwd_scan_dim, True, additional_inputs
+                ),
+                num_leaves_init,
+            )
+
+            # g_xs may contain the gradients for the additional inputs as well which need to be separated
+            new_g_additional_inputs = g_xs[num_leaves_xs:] + old_g_additional_inputs_T
+            g_xs = g_xs[:num_leaves_xs]
+            g_xs = prepare_final_gradients_xs(g_xs, dim, reverse)
+
+        return *[None] * 6, *g_init, *g_xs, *new_g_additional_inputs
```

Comment:
Do we have tests for testing gradient of additional_inputs? I feel we need to modify the backward graph as well:
```
def new_joint_graph(carried_grad_additional_inputs, ...):
   ...grad_additioal_inptus... = joint_graph(...)
   carry (..., grad_addtioanl_inputs + carried_grad_additional_inputs),
```
The init should be a torch.zeros() note that: the joint_graph doesn't take carried_grad_additional_inputs as carry. it only takes additioanl_inputs as part of xs. It's the wrapper new_joint_graph that takes a new carry.

Reply:
I've included a wrapper code that takes care of the gradients for the additional inputs